File: examples/quick-git/git/repository.test.ts
import { afterAll, beforeAll, describe, expect, test } from "bun:test";
import { mkdtemp, rm, writeFile, mkdir } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";

import { GitError, GitRunner } from "./process.ts";
import { Repository } from "./repository.ts";
import { stagedChanges, unstagedChanges } from "./status.ts";

const runner = new GitRunner({ concurrency: 2 });
let root = "";
let repo: Repository;

async function git(...args: string[]): Promise<string> {
  return runner.text(args, { cwd: root });
}

beforeAll(async () => {
  root = await mkdtemp(join(tmpdir(), "quick-git-"));
  await git("init", "-q", "-b", "main");
  await git("config", "user.email", "test@example.com");
  await git("config", "user.name", "Quick Git Tests");
  await git("config", "commit.gpgsign", "false");
  repo = await Repository.open(runner, root);
});

afterAll(async () => {
  await rm(root, { recursive: true, force: true });
});

describe("Repository", () => {
  test("opens the repository from a subdirectory and reports an unborn branch", async () => {
    await mkdir(join(root, "src"), { recursive: true });
    const nested = await Repository.open(runner, join(root, "src"));
    expect(nested.root).toBe(await realRoot());
    const status = await repo.status();
    expect(status.headSha).toBeUndefined();
    expect(status.branch).toBe("main");
    expect(await repo.hasHead()).toBe(false);
    expect(await repo.log()).toEqual([]);
  });

  test("stages, unstages, previews untracked files, and commits", async () => {
    await writeFile(join(root, "src", "app.ts"), `${Array.from({ length: 20 }, (_, index) => `line ${index + 1}`).join("\n")}\n`);
    await writeFile(join(root, "with space.md"), "# Notes\n");
    let status = await repo.status();
    expect(unstagedChanges(status).map((item) => item.path).sort()).toEqual(["src/app.ts", "with space.md"]);

    const preview = await repo.diffUntracked("with space.md");
    expect(preview.kind).toBe("added");
    expect(preview.hunks[0]!.lines.map((line) => line.text)).toEqual(["# Notes"]);

    await repo.stage(["src/app.ts", "with space.md"]);
    status = await repo.status();
    expect(stagedChanges(status).map((item) => item.path).sort()).toEqual(["src/app.ts", "with space.md"]);
    expect(unstagedChanges(status)).toEqual([]);

    // Unstaging on an unborn branch works too.
    await repo.unstage(["with space.md"]);
    status = await repo.status();
    expect(stagedChanges(status).map((item) => item.path)).toEqual(["src/app.ts"]);
    await repo.stage(["with space.md"]);

    await repo.commit("Initial commit\n\nWith a body.");
    status = await repo.status();
    expect(status.headSha).toBeDefined();
    expect(status.entries).toEqual([]);
    const log = await repo.log();
    expect(log).toHaveLength(1);
    expect(log[0]).toMatchObject({ subject: "Initial commit", body: "With a body.", parents: [] });
    expect(log[0]!.refs.some((ref) => ref.kind === "branch" && ref.name === "main" && ref.current)).toBe(true);
    expect(await repo.recentSubjects(5)).toEqual(["Initial commit"]);
    expect(await repo.config("user.name")).toBe("Quick Git Tests");
    expect(await repo.config("does.not.exist")).toBeUndefined();
  });

  test("stages and unstages single hunks and lines through generated patches", async () => {
    const lines = Array.from({ length: 20 }, (_, index) => `line ${index + 1}`);
    lines[0] = "line one!";
    lines.splice(10, 0, "x", "x", "x");
    lines.push("tail added");
    await writeFile(join(root, "src", "app.ts"), `${lines.join("\n")}\n`);
    // Three hunks: the top edit, the inserted x lines, and the tail line.
    let files = await repo.diffWorkingTree("src/app.ts");
    expect(files).toHaveLength(1);
    const file = files[0]!;
    expect(file.hunks).toHaveLength(3);
    const hunkWith = (candidate: typeof file, text: string) =>
      candidate.hunks.findIndex((hunk) => hunk.lines.some((line) => line.kind === "added" && line.text === text));

    // Stage only the tail hunk.
    await repo.stagePatch(file, [{ hunkIndex: hunkWith(file, "tail added") }]);
    let staged = await repo.diffIndex("src/app.ts", undefined);
    expect(staged[0]!.hunks).toHaveLength(1);
    expect(staged[0]!.hunks[0]!.lines.filter((line) => line.kind === "added").map((line) => line.text)).toEqual([
      "tail added",
    ]);
    let unstagedFiles = await repo.diffWorkingTree("src/app.ts");
    expect(unstagedFiles[0]!.hunks).toHaveLength(2);
    expect(hunkWith(unstagedFiles[0]!, "tail added")).toBe(-1);

    // Stage one line pair of the top hunk: the `line 1` replacement only.
    const remaining = unstagedFiles[0]!;
    const topIndex = hunkWith(remaining, "line one!");
    const hunk = remaining.hunks[topIndex]!;
    const lineOneRemoved = hunk.lines.findIndex((line) => line.kind === "removed" && line.text === "line 1");
    const lineOneAdded = hunk.lines.findIndex((line) => line.kind === "added" && line.text === "line one!");
    expect(lineOneRemoved).toBeGreaterThanOrEqual(0);
    expect(lineOneAdded).toBeGreaterThanOrEqual(0);
    await repo.stagePatch(remaining, [{ hunkIndex: topIndex, lines: new Set([lineOneRemoved, lineOneAdded]) }]);
    staged = await repo.diffIndex("src/app.ts", undefined);
    const stagedText = staged[0]!.hunks.flatMap((hunk) => hunk.lines.filter((line) => line.kind !== "context").map((line) => `${line.kind}:${line.text}`));
    expect(stagedText).toEqual(["removed:line 1", "added:line one!", "added:tail added"]);

    // Unstage the tail hunk again from the staged diff.
    const tailHunkIndex = staged[0]!.hunks.findIndex((hunk) => hunk.lines.some((line) => line.text === "tail added"));
    await repo.unstagePatch(staged[0]!, [{ hunkIndex: tailHunkIndex }]);
    staged = await repo.diffIndex("src/app.ts", undefined);
    expect(staged[0]!.hunks.flatMap((hunk) => hunk.lines.filter((line) => line.kind === "added").map((line) => line.text))).toEqual(["line one!"]);

    // Discard the x-lines hunk from the working tree; the file keeps the staged edit and the tail.
    unstagedFiles = await repo.diffWorkingTree("src/app.ts");
    const xHunk = unstagedFiles[0]!.hunks.findIndex((hunk) => hunk.lines.some((line) => line.kind === "added" && line.text === "x"));
    expect(xHunk).toBeGreaterThanOrEqual(0);
    await repo.discardPatch(unstagedFiles[0]!, [{ hunkIndex: xHunk }]);
    const content = await Bun.file(join(root, "src", "app.ts")).text();
    expect(content).not.toContain("\nx\n");
    expect(content).toContain("line one!\n");
    expect(content).toContain("tail added\n");
    await repo.unstageAll();
    files = await repo.diffWorkingTree("src/app.ts");
    expect(files[0]!.hunks.length).toBeGreaterThan(0);
  });

  test("discards tracked changes and trashes untracked files through the injected remover", async () => {
    const trashed: string[] = [];
    const custom = await Repository.open(runner, root, {
      trash: async (path) => {
        trashed.push(path);
        await rm(path, { force: true });
      },
    });
    await writeFile(join(root, "junk.txt"), "junk\n");
    await custom.discard(["src/app.ts"], ["junk.txt"]);
    expect(trashed).toEqual([join(await realRoot(), "junk.txt")]);
    const status = await repo.status();
    expect(status.entries).toEqual([]);
  });

  test("manages branches, stashes, and worktrees", async () => {
    await repo.createBranch("feature/login", { checkout: false });
    let refs = await repo.refs();
    expect(refs.local.map((branch) => branch.name).sort()).toEqual(["feature/login", "main"]);
    expect(refs.local.find((branch) => branch.name === "main")?.current).toBe(true);

    await writeFile(join(root, "stash-me.txt"), "wip\n");
    await repo.stashPush({ message: "my stash", includeUntracked: true });
    const stashes = await repo.stashes();
    expect(stashes).toHaveLength(1);
    expect(stashes[0]).toMatchObject({ ref: "stash@{0}", branch: "main", summary: "my stash" });
    await repo.stashPop("stash@{0}");
    expect((await repo.status()).entries.map((entry) => entry.path)).toEqual(["stash-me.txt"]);
    await rm(join(root, "stash-me.txt"));

    const worktreePath = join(root, "..", `${root.split("/").pop()}-feature`);
    await repo.addWorktree({ path: worktreePath, branch: "feature/login" });
    let worktrees = await repo.worktrees();
    expect(worktrees.map((worktree) => [worktree.branchName, worktree.main])).toEqual([
      ["main", true],
      ["feature/login", false],
    ]);
    refs = await repo.refs();
    expect(refs.local.find((branch) => branch.name === "feature/login")?.worktreePath).toBe(worktrees[1]!.path);

    const linked = await repo.worktree(worktreePath);
    expect(linked.info.commonDir).toBe(repo.info.commonDir);
    await writeFile(join(worktreePath, "feature.txt"), "feature\n");
    await linked.stage(["feature.txt"]);
    await linked.commit("Feature work");
    expect((await linked.log())[0]!.subject).toBe("Feature work");
    expect((await repo.log())[0]!.subject).toBe("Initial commit");

    await repo.removeWorktree(worktreePath, true);
    worktrees = await repo.worktrees();
    expect(worktrees).toHaveLength(1);
    await repo.deleteBranch("feature/login", true);
    refs = await repo.refs();
    expect(refs.local.map((branch) => branch.name)).toEqual(["main"]);
  });

  test("reports commit files and diffs against the first parent", async () => {
    const [head] = await repo.log({ limit: 1 });
    const files = await repo.commitFiles(head!.sha);
    expect(files.map((file) => [file.status, file.path]).sort()).toEqual([
      ["A", "src/app.ts"],
      ["A", "with space.md"],
    ]);
    const diff = await repo.diffCommit(head!.sha, "src/app.ts");
    expect(diff).toHaveLength(1);
    expect(diff[0]!.kind).toBe("added");
  });

  test("surfaces git failures as typed errors and honours cancellation", async () => {
    const failure = await repo.switchBranch("does-not-exist").catch((error: unknown) => error);
    expect(failure).toBeInstanceOf(GitError);
    expect((failure as GitError).exitCode).not.toBe(0);
    expect((failure as GitError).summary.length).toBeGreaterThan(0);

    const controller = new AbortController();
    controller.abort();
    const cancelled = await repo.status({ signal: controller.signal }).catch((error: unknown) => error);
    expect(cancelled).toBeInstanceOf(GitError);
    expect((cancelled as GitError).aborted).toBe(true);
  });
});

async function realRoot(): Promise<string> {
  return (await runner.text(["rev-parse", "--show-toplevel"], { cwd: root })).trim();
}

File: examples/quick-git/internal/git/repository_test.go
package git

import (
	"context"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
)

func testRepository(t *testing.T) (*Repository, context.Context) {
	t.Helper()
	root := t.TempDir()
	// Keep user signing, hooks, aliases, and global exclusions out of fixtures.
	t.Setenv("GIT_CONFIG_GLOBAL", os.DevNull)
	t.Setenv("GIT_CONFIG_NOSYSTEM", "1")
	runner := NewRunner(2)
	ctx := context.Background()
	for _, args := range [][]string{
		{"init", "-q", "-b", "main"},
		{"config", "user.email", "test@example.com"},
		{"config", "user.name", "Quick Git Tests"},
		{"config", "commit.gpgsign", "false"},
		{"config", "core.hooksPath", filepath.Join(root, "no-hooks")},
	} {
		_, err := runner.Run(ctx, args, CommandOptions{Cwd: root})
		requireOK(t, err)
	}
	repo, err := Open(ctx, runner, root, nil)
	requireOK(t, err)
	return repo, ctx
}

func requireOK(t *testing.T, err error) {
	t.Helper()
	if err != nil {
		t.Fatal(err)
	}
}

func writeRepoFile(t *testing.T, repo *Repository, name, content string) {
	t.Helper()
	path := filepath.Join(repo.Root(), name)
	requireOK(t, os.MkdirAll(filepath.Dir(path), 0o755))
	requireOK(t, os.WriteFile(path, []byte(content), 0o644))
}

func repoText(t *testing.T, repo *Repository, args ...string) string {
	t.Helper()
	text, err := repo.Runner.Text(context.Background(), args, CommandOptions{Cwd: repo.Root()})
	requireOK(t, err)
	return text
}

func TestRepositoryUnbornStageCommitAndHistory(t *testing.T) {
	repo, ctx := testRepository(t)
	writeRepoFile(t, repo, "src/app.go", "first\nsecond\n")
	writeRepoFile(t, repo, "with space.md", "# Notes\n")
	nested, err := Open(ctx, repo.Runner, filepath.Join(repo.Root(), "src"), nil)
	requireOK(t, err)
	if nested.Root() != repo.Root() {
		t.Fatalf("nested root = %q", nested.Root())
	}
	status, err := repo.Status(ctx)
	requireOK(t, err)
	if status.Branch != "main" || status.HeadSha != "" || len(UnstagedChanges(status)) != 2 {
		t.Fatalf("unborn status: %+v", status)
	}
	head, err := repo.HasHead(ctx)
	requireOK(t, err)
	if head {
		t.Fatal("unborn repository has HEAD")
	}
	commits, err := repo.Log(ctx, LogOptions{})
	requireOK(t, err)
	if len(commits) != 0 {
		t.Fatal(commits)
	}
	preview, err := repo.DiffUntracked(ctx, "with space.md")
	requireOK(t, err)
	if len(preview.Files) != 1 || preview.Files[0].Kind != FileAdded || preview.Added != 1 {
		t.Fatalf("untracked preview: %+v", preview)
	}
	requireOK(t, repo.Stage(ctx, []string{"src/app.go", "with space.md"}))
	requireOK(t, repo.Unstage(ctx, []string{"with space.md"}))
	status, err = repo.Status(ctx)
	requireOK(t, err)
	if got := StagedChanges(status); len(got) != 1 || got[0].Path != "src/app.go" {
		t.Fatal(got)
	}
	requireOK(t, repo.StageAll(ctx))
	requireOK(t, repo.Commit(ctx, "Initial commit\n\nWith a body.\n", CommitOptions{}))
	commits, err = repo.Log(ctx, LogOptions{})
	requireOK(t, err)
	if len(commits) != 1 || commits[0].Subject != "Initial commit" || strings.TrimSpace(commits[0].Body) != "With a body." || len(commits[0].Parents) != 0 {
		t.Fatalf("log: %+v", commits)
	}
	files, err := repo.CommitFiles(ctx, commits[0].Sha)
	requireOK(t, err)
	if len(files) != 2 {
		t.Fatalf("commit files: %+v", files)
	}
	diff, err := repo.DiffCommit(ctx, commits[0].Sha, "")
	requireOK(t, err)
	if len(diff.Files) != 2 || diff.Added != 3 {
		t.Fatalf("multi-file diff: %+v", diff)
	}
	status, err = repo.Status(ctx)
	requireOK(t, err)
	if !IsClean(status) {
		t.Fatal(status)
	}
}

func changes(file DiffFile) []string {
	var result []string
	for _, hunk := range file.Hunks {
		for _, line := range hunk.Lines {
			if line.Kind != LineContext {
				result = append(result, string(line.Kind)+":"+line.Text)
			}
		}
	}
	return result
}

func findHunk(t *testing.T, file DiffFile, text string) int {
	t.Helper()
	for i, hunk := range file.Hunks {
		for _, line := range hunk.Lines {
			if line.Text == text && line.Kind != LineContext {
				return i
			}
		}
	}
	t.Fatalf("missing changed line %q in %+v", text, file)
	return -1
}

func TestRepositoryPartialStageUnstageDiscardWithOffsets(t *testing.T) {
	repo, ctx := testRepository(t)
	var lines []string
	for i := 1; i <= 30; i++ {
		lines = append(lines, fmt.Sprintf("line %d", i))
	}
	writeRepoFile(t, repo, "file.txt", strings.Join(lines, "\n")+"\n")
	requireOK(t, repo.StageAll(ctx))
	requireOK(t, repo.Commit(ctx, "Base", CommitOptions{}))
	lines[0] = "line one!"
	lines = append(append(append([]string{}, lines[:12]...), "x", "x", "x"), lines[12:]...)
	lines = append(lines, "tail added")
	writeRepoFile(t, repo, "file.txt", strings.Join(lines, "\n")+"\n")
	diff, err := repo.DiffWorkingTree(ctx, "file.txt")
	requireOK(t, err)
	file := diff.Files[0]
	if len(file.Hunks) != 3 {
		t.Fatalf("hunks: %+v", file.Hunks)
	}
	requireOK(t, repo.StagePatch(ctx, file, []HunkSelection{{HunkIndex: findHunk(t, file, "tail added")}}))
	staged, err := repo.DiffIndex(ctx, "file.txt", "")
	requireOK(t, err)
	if got := changes(staged.Files[0]); !reflect.DeepEqual(got, []string{"added:tail added"}) {
		t.Fatal(got)
	}
	diff, err = repo.DiffWorkingTree(ctx, "file.txt")
	requireOK(t, err)
	file = diff.Files[0]
	top := findHunk(t, file, "line one!")
	selection := HunkSelection{HunkIndex: top, Lines: map[int]struct{}{}}
	for i, line := range file.Hunks[top].Lines {
		if line.Text == "line one!" || line.Text == "line 1" {
			selection.Lines[i] = struct{}{}
		}
	}
	requireOK(t, repo.StagePatch(ctx, file, []HunkSelection{selection}))
	staged, err = repo.DiffIndex(ctx, "file.txt", "")
	requireOK(t, err)
	if got := changes(staged.Files[0]); !reflect.DeepEqual(got, []string{"removed:line 1", "added:line one!", "added:tail added"}) {
		t.Fatal(got)
	}
	requireOK(t, repo.UnstagePatch(ctx, staged.Files[0], []HunkSelection{{HunkIndex: findHunk(t, staged.Files[0], "tail added")}}))
	if got := repoText(t, repo, "show", ":file.txt"); strings.Contains(got, "tail added") || !strings.Contains(got, "line one!") {
		t.Fatal(got)
	}
	diff, err = repo.DiffWorkingTree(ctx, "file.txt")
	requireOK(t, err)
	requireOK(t, repo.DiscardPatch(ctx, diff.Files[0], []HunkSelection{{HunkIndex: findHunk(t, diff.Files[0], "x")}}))
	content, err := os.ReadFile(filepath.Join(repo.Root(), "file.txt"))
	requireOK(t, err)
	if strings.Contains(string(content), "\nx\n") || !strings.Contains(string(content), "tail added") || !strings.Contains(string(content), "line one!") {
		t.Fatal(string(content))
	}
}

func TestRepositoryPartialNewAndDeletedFileAndNoNewline(t *testing.T) {
	for _, kind := range []string{"added", "deleted", "no-newline"} {
		t.Run(kind, func(t *testing.T) {
			repo, ctx := testRepository(t)
			requireOK(t, repo.Commit(ctx, "Base", CommitOptions{AllowEmpty: true}))
			switch kind {
			case "added":
				writeRepoFile(t, repo, "new.txt", "one\ntwo\nthree\n")
				requireOK(t, repo.StageAll(ctx))
				diff, err := repo.DiffIndex(ctx, "new.txt", "")
				requireOK(t, err)
				requireOK(t, repo.UnstagePatch(ctx, diff.Files[0], []HunkSelection{{HunkIndex: 0, Lines: map[int]struct{}{1: {}}}}))
				if got := repoText(t, repo, "show", ":new.txt"); got != "one\nthree\n" {
					t.Fatal(got)
				}
			case "deleted":
				writeRepoFile(t, repo, "old.txt", "one\ntwo\nthree\n")
				requireOK(t, repo.StageAll(ctx))
				requireOK(t, repo.Commit(ctx, "Add old file", CommitOptions{}))
				requireOK(t, os.Remove(filepath.Join(repo.Root(), "old.txt")))
				diff, err := repo.DiffWorkingTree(ctx, "old.txt")
				requireOK(t, err)
				requireOK(t, repo.StagePatch(ctx, diff.Files[0], []HunkSelection{{HunkIndex: 0, Lines: map[int]struct{}{1: {}}}}))
				if got := repoText(t, repo, "show", ":old.txt"); got != "one\nthree\n" {
					t.Fatal(got)
				}
			case "no-newline":
				writeRepoFile(t, repo, "plain.txt", "before")
				requireOK(t, repo.StageAll(ctx))
				requireOK(t, repo.Commit(ctx, "Before", CommitOptions{}))
				writeRepoFile(t, repo, "plain.txt", "after")
				diff, err := repo.DiffWorkingTree(ctx, "plain.txt")
				requireOK(t, err)
				requireOK(t, repo.StagePatch(ctx, diff.Files[0], []HunkSelection{{HunkIndex: 0}}))
				if got := repoText(t, repo, "show", ":plain.txt"); got != "after" {
					t.Fatal(got)
				}
			}
		})
	}
}

func TestRepositoryBranchesStashesWorktreesAndErrors(t *testing.T) {
	repo, ctx := testRepository(t)
	writeRepoFile(t, repo, "base.txt", "base\n")
	requireOK(t, repo.StageAll(ctx))
	requireOK(t, repo.Commit(ctx, "Initial", CommitOptions{}))
	requireOK(t, repo.CreateBranch(ctx, "feature/login", "", false))
	refs, err := repo.Refs(ctx)
	requireOK(t, err)
	if len(refs.Local) != 2 {
		t.Fatal(refs)
	}
	writeRepoFile(t, repo, "stash.txt", "wip\n")
	requireOK(t, repo.StashPush(ctx, "my stash", true, false))
	stashes, err := repo.Stashes(ctx)
	requireOK(t, err)
	if len(stashes) != 1 || stashes[0].Summary != "my stash" || stashes[0].Branch != "main" {
		t.Fatal(stashes)
	}
	requireOK(t, repo.StashPop(ctx, stashes[0].Ref))
	requireOK(t, os.Remove(filepath.Join(repo.Root(), "stash.txt")))
	linkedPath := filepath.Join(t.TempDir(), "feature")
	requireOK(t, repo.AddWorktree(ctx, linkedPath, "", "feature/login", ""))
	linked, err := repo.Worktree(ctx, linkedPath)
	requireOK(t, err)
	if linked.Info.CommonDir != repo.Info.CommonDir {
		t.Fatalf("common dirs: %q != %q", linked.Info.CommonDir, repo.Info.CommonDir)
	}
	writeRepoFile(t, linked, "feature.txt", "feature\n")
	requireOK(t, linked.StageAll(ctx))
	requireOK(t, linked.Commit(ctx, "Feature work", CommitOptions{}))
	if got := repoText(t, repo, "log", "-1", "--format=%s"); got != "Initial\n" {
		t.Fatal(got)
	}
	worktrees, err := repo.Worktrees(ctx)
	requireOK(t, err)
	if len(worktrees) != 2 {
		t.Fatal(worktrees)
	}
	requireOK(t, repo.RemoveWorktree(ctx, linkedPath, true))
	requireOK(t, repo.DeleteBranch(ctx, "feature/login", true))
	writeRepoFile(t, repo, "base.txt", "changed\n")
	writeRepoFile(t, repo, "junk.txt", "junk\n")
	var trashed []string
	repo.Trash = func(path string) error { trashed = append(trashed, path); return os.Remove(path) }
	requireOK(t, repo.Discard(ctx, []string{"base.txt"}, []string{"junk.txt"}))
	if !reflect.DeepEqual(trashed, []string{filepath.Join(repo.Root(), "junk.txt")}) {
		t.Fatal(trashed)
	}
	var failure *Error
	if err := repo.SwitchBranch(ctx, "does-not-exist"); !errors.As(err, &failure) || failure.ExitCode == 0 || failure.Summary() == "" {
		t.Fatalf("failure: %v", err)
	}
	cancelled, cancel := context.WithCancel(ctx)
	cancel()
	if _, err := repo.Status(cancelled); !errors.As(err, &failure) || !failure.Aborted {
		t.Fatalf("cancel: %v", err)
	}
}
